pkg/controller/machineset/controller.go

```diff
@@ -129,7 +131,22 @@ func (c *MachineSetControllerImpl) Reconcile(machineSet *v1alpha1.MachineSet) er
 				continue
 			}
 		}
-		filteredMachines = append(filteredMachines, machine)
+		machineNames = append(machineNames, machine.Name)
+		machineSetMachines[machine.Name] = machine
+	}
+
+	// Sort the machines deterministically to delete machines in the same order.
+	// In general the machines are sequenced randomly depending on Machines().List() result.
+	// When number of replicas goes down, the reconcilication loop is called multiple
+	// times, each time taking a different subset of machines to be deleted.
+	// Which may result in deletion of more machines than requested (followed
+	// by recreation of new machines). Given each machine represents a node,
+	// deleting more nodes than expected can be very disruptive.
+	sort.Strings(machineNames)
```

Comment:
So far the machines were listed randomly. Given a machine name is generated randomly, the original deletion policy (deleting first machines in a list) is preserved.

Reply:
> Given a machine name is generated randomly, the original deletion policy (deleting first machines in a list) is preserved.

How does the final sort preserve the original deletion policy?

Reply:
The original deletion policy takes first items from a list where whose order is generated randomly. Reordering the items based on a name that is generated randomly as well changes order if items from one random sequence to another one.